src/gwt/src/org/rstudio/core/client/VirtualConsole.java

```diff
@@ -691,8 +752,72 @@ public void submit(String data, String clazz, boolean forceNewRange, boolean ari
+               {
+                  tail += 1;
+                  break;
+               }
+               
+               // check for an escape forcing a new span
+               if (parent_ != null)
+               {
+                  Pattern groupStartPattern = Pattern.create("^\\033G(\\d+);", "m");
+                  Match groupStartMatch = groupStartPattern.match(data.substring(head), 0);
+                  if (groupStartMatch != null)
+                  {
+                     String type = groupStartMatch.getGroup(1);
+                     String groupClazz = groupTypeToClazz(type);
+                     
+                     // re-use the previous group if we're closing and re-opening
+                     // a group of the same type
+                     Node lastNode = parent_.getLastChild();
+                     if (Element.is(lastNode))
+                     {
+                        Element lastEl = Element.as(lastNode);
+                        if (lastEl.hasClassName(groupClazz))
+                        {
+                           parent_ = lastEl;
+                           tail += groupStartMatch.getValue().length() - 1;
+                           break;
+                        }
+                     }
+                     
+                     // otherwise, create a new group span and use it
+                     SpanElement spanEl = Document.get().createSpanElement();
+                     spanEl.addClassName(RES.styles().group());
+                     spanEl.addClassName(groupClazz);
+                     parent_.appendChild(spanEl);
+                     parent_ = spanEl;
+                     tail += groupStartMatch.getValue().length() - 1;
+                     break;
+                  }
+                  
+                  Pattern groupEndPattern = Pattern.create("^\\033g", "m");
+                  Match groupEndMatch = groupEndPattern.match(data.substring(head), 0);
+                  if (groupEndMatch != null)
+                  {
+                     parent_ = parent_.getParentElement();
+                     tail += groupEndMatch.getValue().length() - 1;
+                     break;
+                  }
+               }
+               
+               // check for embedded custom highlight rules
+               Pattern customPattern = Pattern.create("^\\033H(\\d+);([^]*?)\\033h", "m");
+               Match customMatch = customPattern.match(data.substring(head), 0);
+               if (customMatch != null)
+               {
+                  String type = customMatch.getGroup(1);
+                  String code = customMatch.getGroup(2);
+                  text(code, typeToClazz(type), type != "3");
+                  tail = head + customMatch.getValue().length();
+                  break;
+               }
+               
                // match complete CSI codes
-               Match csiMatch = AnsiCode.CSI_PATTERN.match(data, pos);
```

Comment:
I realized that we weren't actually anchoring any of our ANSI escape matchers to the position of the escape we were processing, so you'll see a number of changes of this form (since, IIUC, we want to match _only_ at the current position we're parsing)